### ISSUE TITLE
Support incremental compilation

### ISSUE THREAD

**Issue by nbirillo** [JetBrains-Research/reflekt#69]
Currently, Reflekt does not support the incremental compilation process. We have to get all compiled instances per module and merge old results with new ones

**Comment by nbirillo** [comment]
Let me tell about the current situation with this issue.

Incremental compilation is not directly a compiler API. This means that currently, the compiler API does not provide an opportunity to interfere with the incremental compilation process, namely to do something with files that have not actually changed but must be recompiled in the plugin.

Consider an example, since we are actually replacing the Reflekt queries and replacing reflection to IR generation, then we do not contain Reflekt calls in the compiled code. Let's say, before changing the project, we were looking for classes with annotation B, and after the first round of compilation, we added another class C with annotation B to the project. But since the IR for the first Reflekt call has already been replaced, and in fact, the file with the Reflkt call itself has not changed, we do not have its access during the second round of the compilation process. In the result, we will not change the replaced IR and the result will be incorrect since it will not contain the new class C.

We tried to find weak points to change this, but unfortunately, this is not possible at the present time.
However, this functionality can be useful not only in Reflekt, but also to other plugin developers. Therefore, the ability to interfere with the incremental compilation process from a compiler plugin will be added in the next Kotlin releases. 
At the moment, we have plans to add the necessary changes in Kotlin 1.7.0 (firstly for Gradle and JVM target, next, support for JPS and other targets like JS).

To summarize, we are doing everything possible to make Reflekt available for use with incremental compilation, but we cannot support it before version 1.7.0 (or maybe a little later)
